### ISSUE TITLE
Why does this even compile? (Circular self-initialisation).

### ISSUE THREAD

**Issue by ericmutta** [dotnet/vblang#325]
After doing some "Find & Replace" in some code I was working on I wound up with code like this:

```vb
Public Class TestClass
  Private ReadOnly SomeFoo As SomeOtherClass = Me.SomeFoo
End Class
```

Strangely enough, that code compiles without a problem and when you try to access the `SomeFoo` field at run-time you get a null reference exception. It's a little hard to spot at first, but the problem is `SomeFoo` is being initialised to the value of `SomeFoo`. **It is being initialised with its own value** which makes no sense logically because it is circular.

The compiler already rejects other circular scenarios such as a class inheriting from itself:

![vs_self_init](https://user-images.githubusercontent.com/20465797/42424128-2875fea0-830f-11e8-80e7-2f3398e834cd.png)

...so is there any reason why circular initialisation of a field is allowed (or even useful)?

**Comment by KathleenDollard** [comment]
@ericmutta Can you file this as an [issue on Roslyn](https://github.com/dotnet/roslyn/issues)

On first look, it seems to be a bug.

**Comment by jmarolf** [comment]
@ericmutta I agree with @KathleenDollard that this is a bug on roslyn.  We should likely provide a compiler warning or error here.

regarding "Is is ever useful for a variable to reference itself during declaration", VB allows this so lambdas can be recursive:
```VB
Dim factorial As Func(Of Integer, Integer) = Function(n As Integer) As Integer
                                                 If n <= 1 Then
                                                     Return 1
                                                 Else
                                                     Return factorial(n - 1) * n
                                                 End If
                                             End Function
```

**Comment by ericmutta** [comment]
@KathleenDollard I filed the bug at [#28398](https://github.com/dotnet/roslyn/issues/28398) :+1: 

@jmarolf thanks for the pointer about recursive lambdas! Never in a million years would that have occurred to me as a valid use of this behaviour :smile:

**Comment by AdamSpeight2008** [comment]
I think they'll (Roslyn) that this isn't a bug and is be design. It is a good candidate for an analyser though.

**Comment by jrmoreno1** [comment]
It's not a bug because a variable, member or property can be used as part of an expression that modifies the variable, member or property.  x  = x +1 works for all of them.  No reason to make initialization special, particularly since you can come up with scenarios where it makes sense for member and properties (and even local variables as the lambda example shows).  This is really, if it hurts don't do that situation.

**Comment by reduckted** [comment]
This is a bit different to something like `x = x + 1`. In this case, there's a _read-only_ field being _initialized_ to itself.

Regardless, if you try this in C# you get a compiler error:

```c#
class TestClass {
    private readonly TestClass Foo = Foo;
                                     ~~~ // ERROR: A field initializer cannot reference the non-static field, method, or property 'TestClass.Foo'
}
```

What's interesting is that C# _does_ let you do the same thing inside the constructor, though it does show a warning:

```c#
class TestClass {
    private readonly TestClass Foo;

    TestClass() {
        Foo = Foo;
        ~~~~~~~~~ // WARNING: Assignment made to same variable; did you mean to assign something else?
    }
}
```

**Comment by AnthonyDGreen** [comment]
It looks like a good candidate for warning waves but I understand why it's not illegal. You might also want to check for circular initialization of readonly auto props, but there are situations in which that won't be wrong so... maybe not.
